### PR TITLE
[megatron] feat: support DeepSeek V4 GRPO

### PR DESCRIPTION
**Pull request by HollowMan6**
### What does this PR do?

Adds DeepSeek V4 Flash GRPO support with Megatron-Bridge actor/ref workers, vLLM rollout, FP8/MXFP4 weight transfer handling, and checkpoint save/export verification.

Need
- https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3969

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: 
   - https://github.com/verl-project/verl/pulls?q=is%3Apr+is%3Aopen+%22DeepSeek+V4+Flash%22                                                                                                  
   - https://github.com/verl-project/verl/pulls?q=is%3Apr+is%3Aopen+vllm+fp8+mxfp4                                                                                                           
   - https://github.com/verl-project/verl/pulls?q=is%3Apr+is%3Aopen+Megatron+Bridge+DeepSeek
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

End-to-end verification:                                                                                                                                                                    
                                                                                                                                                                                                                                                                                                                                                                 
<img width="1696" height="246" alt="img_v3_0213j_ad93d239-e0af-4349-a545-6cf37f23a07g" src="https://github.com/user-attachments/assets/5cecfc6d-dd7e-4724-aa7a-0528e5b0229b" />

### API and Usage Example

This PR adds an example script:

bash examples/grpo_trainer/run_deepseek_v4_flash_megatron.sh

### Design & Code Changes

- Add DeepSeek V4 Flash GRPO example script and README entry.
- Add vLLM FP8/MXFP4 utilities for quantized rollout weight preparation and reload-safe parameter handling.
- Add bucketed Megatron-to-vLLM weight transfer coverage for quantized weights.
- Normalize DeepSeek V4 vLLM HF overrides, including MTP disablement and Yarn RoPE scaling fields.
- Adjust Megatron transformer config handling for disabled MTP layers.
- Add model config support needed by DeepSeek V4 Flash.
- Add unit coverage for vLLM FP8 utilities, bucketed transfer behavior, and model config handling.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [X] Read the [Contribute Guide](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md).
- [X] Apply [pre-commit checks](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [X] Add / Update [the documentation](https://github.com/verl-project/verl/tree/main/docs).
- [X] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [X] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [X] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

File: verl/workers/engine/megatron/transformer_impl.py [hunk 1]
```diff
@@ -214,6 +214,10 @@ def _build_tf_config(self):
             }
             for key, value in override_transformer_config.items():
                 provider_overrides[key] = value
+            if not self.model_config.mtp.enable:
+                provider_overrides["mtp_num_layers"] = 0
+                if provider.csa_compress_ratios is not None:
+                    provider_overrides["csa_compress_ratios"] = provider.csa_compress_ratios[: provider.num_layers]
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/workers/engine/megatron/transformer_impl.py]: fixed

File: verl/utils/vllm/vllm_fp8_utils.py [hunk 2]
```diff
@@ -46,46 +54,411 @@ class FP8State:
 fp8_state: FP8State = FP8State()
 
 
-def is_fp8_model(vllm_config):
-    from vllm.model_executor.layers.quantization.fp8 import Fp8Config
+def _copy_param_attributes(dst_param, src_param):
+    base_param_dir = dir(torch.nn.Parameter)
+    for attr in dir(src_param):
+        if attr not in base_param_dir and not attr.startswith("__"):
+            setattr(dst_param, attr, getattr(src_param, attr))
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/utils/vllm/vllm_fp8_utils.py]: fixed

File: verl/workers/rollout/vllm_rollout/vllm_async_server.py [hunk 3]
```diff
@@ -833,6 +833,22 @@ def _apply_quantization(self) -> tuple[Optional[str], dict]:
...
+        rope_scaling = getattr(hf_config, "rope_scaling", None)
+        if isinstance(rope_scaling, dict):
+            rope_type = rope_scaling.get("rope_type") or rope_scaling.get("type")
+            if rope_type == "yarn":
+                rope_scaling_override = dict(rope_scaling)
+                rope_scaling_override["rope_type"] = rope_type
+                rope_scaling_override["type"] = rope_type
+                for key in ("factor", "beta_fast", "beta_slow"):
+                    if key in rope_scaling_override:
+                        rope_scaling_override[key] = float(rope_scaling_override[key])
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/workers/rollout/vllm_rollout/vllm_async_server.py]: fixed

File: verl/utils/vllm/vllm_fp8_utils.py [hunk 4]
```diff
@@ -46,46 +54,414 @@ class FP8State:
...
-def is_fp8_model(vllm_config):
-    from vllm.model_executor.layers.quantization.fp8 import Fp8Config
+def _copy_param_attributes(dst_param, src_param):
+    base_param_dir = dir(torch.nn.Parameter)
+    for attr in dir(src_param):
+        if attr not in base_param_dir and not attr.startswith("__"):
+            try:
+                setattr(dst_param, attr, getattr(src_param, attr))
+            except AttributeError:
+                pass
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/utils/vllm/vllm_fp8_utils.py]: fixed

File: verl/utils/vllm/vllm_fp8_utils.py [hunk 5]
```diff
@@ -46,46 +54,414 @@ class FP8State:
...
+                candidates.append(module_path_str.replace(old, new, 1).split("."))
+
+        return candidates
+
+    packed_modules_mapping = getattr(model, "packed_modules_mapping", {})
     reversed_mapping = {
         original_name: fused_name
         for fused_name, original_names_list in packed_modules_mapping.items()
         for original_name in original_names_list
     }
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/utils/vllm/vllm_fp8_utils.py]: fixed

File: verl/utils/vllm/vllm_fp8_utils.py [hunk 6]
```diff
@@ -96,21 +472,109 @@ def is_fp8_weight(name, model):
...
+def _model_type(model):
+    for obj in (model, getattr(model, "config", None), getattr(model, "hf_config", None)):
+        if obj is None:
+            continue
+        model_type = getattr(obj, "model_type", None)
+        if model_type is not None:
+            return model_type
+    config = getattr(model, "config", None)
+    text_config = getattr(config, "text_config", None)
+    return getattr(text_config, "model_type", None)
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/utils/vllm/vllm_fp8_utils.py]: fixed

File: verl/workers/rollout/vllm_rollout/vllm_async_server.py [hunk 7]
```diff
@@ -833,6 +833,22 @@ def _apply_quantization(self) -> tuple[Optional[str], dict]:
         """Process quantization config. Returns (quantization_str, hf_overrides)."""
         quantization = self.config.quantization
         hf_overrides = {}
+        hf_config = self.model_config.hf_config
+        if not self.model_config.mtp.enable:
+            for field in ("num_nextn_predict_layers", "mtp_num_hidden_layers"):
+                if hasattr(hf_config, field):
+                    hf_overrides[field] = 0
+        rope_scaling = getattr(hf_config, "rope_scaling", None)
+        if isinstance(rope_scaling, dict):
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/workers/rollout/vllm_rollout/vllm_async_server.py]: fixed

File: verl/workers/rollout/vllm_rollout/bucketed_weight_transfer.py [hunk 8]
```diff
@@ -125,31 +130,37 @@ async def async_send_weights(self, weights):
...
+
                 # fill the tensor bucket
-                if offset + weight.nbytes > self.bucket_size and len(bucket_meta) > 0:
+                if aligned_offset + weight_nbytes > self.bucket_size and len(bucket_meta) > 0:
                     get_torch_device().synchronize()
                     self.socket.send_pyobj({"bucket_meta": bucket_meta, "is_last": False})
                     self.socket.recv()
                     bucket_meta = {}
                     offset = 0
+                    aligned_offset = 0
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/workers/rollout/vllm_rollout/bucketed_weight_transfer.py]: fixed

**Comment by Meirtz** [comment]
Hey @HollowMan6 — opened #6515 as the symmetric fix for the `vanilla_mbridge=True` (ISEEKYAN/mbridge) branch in the same `_build_tf_config` function. The MTP-disable + `csa_compress_ratios` trim pattern is the same as your change here, just applied at the `set_extra_args` rebuild point instead of inside `provider_overrides`.

End-to-end verified on GB200 single GPU through ISEEKYAN/mbridge + DSv4 hybrid attention (forward+backward+optimizer.step), pairing well with your `vanilla_mbridge=False` path coverage. They're complementary and independent — either can merge first.

Also kicked an upstream bridge-side default at NVIDIA-NeMo/Megatron-Bridge#4003 for `dsa_indexer_loss_coeff` (`MLATransformerConfig` defaults the field to `None` so `csa.py`'s `getattr(..., 0.0)` fallback never fires — you're working around it via `++override_transformer_config.dsa_indexer_loss_coeff=0.0` in `run_deepseek_v4_flash_megatron.sh`, that override stays valid but becomes redundant once the bridge sets a non-None default). Mentioning in case useful coordination.

**Comment by PeterYang12** [comment]
@HollowMan6  @wuxibin89  Do we have any plan on this PR? Some of AMD works for DSV4 enabling depend on this PR.

File: verl/utils/vllm/vllm_fp8_utils.py [hunk 9]
```diff
@@ -46,6 +51,242 @@ class FP8State:
 fp8_state: FP8State = FP8State()
 
 
+def _is_deepseek_v4_mega_moe_module(module):
```
Comment (wuxibin89) [inline verl/utils/vllm/vllm_fp8_utils.py]: Could you move dsv4 related function to a separate file: `verl/utils/vllm/vllm_dsv4_fp8_utils.py`?

Comment (HollowMan6) [reply verl/utils/vllm/vllm_fp8_utils.py]: done

File: verl/workers/rollout/vllm_rollout/utils.py [hunk 10]
```diff
@@ -274,14 +275,23 @@ def update_weights_from_ipc(self, peft_config: dict = None, base_sync_done=False
                 patch_vllm_moe_model_weight_loader(model)
 
         assert self.device is not None
+        quant_reload_state = False
+        if is_fp8_model(self.model_runner.vllm_config) and not (peft_config and base_sync_done):
+            from verl.utils.vllm.vllm_fp8_utils import prepare_quanted_weights_for_loading
+
+            quant_reload_state = prepare_quanted_weights_for_loading(self.model_runner)
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/workers/rollout/vllm_rollout/utils.py:282]: use_standard_weight_load already excluded FP8 at the base commit. Non-DSV4 FP8 still follows the existing per-bucket load_quanted_weights path. Enabling standard postprocessing afterward would be a new, potentially non-idempotent behavior, not a regression fix; the suggested assignment also occurs too late to run the corresponding pre-load setup.

File: verl/utils/vllm/vllm_dsv4_fp8_utils.py [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+def _is_mega_moe_module(module):
+    from vllm.models.deepseek_v4.nvidia.model import DeepseekV4MegaMoEExperts
+
+    return isinstance(module, DeepseekV4MegaMoEExperts)
+
+
+def _is_mxfp4_fused_moe_module(module):
+    from vllm.model_executor.layers.quantization.mxfp4 import Mxfp4MoEMethod
+
+    return isinstance(module, FusedMoE) and isinstance(module.quant_method, Mxfp4MoEMethod)
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/utils/vllm/vllm_dsv4_fp8_utils.py:51]: The DSV4 backend imports are lazy and execute only after the model is identified as deepseek_v4. Missing DSV4 backend modules should fail closed. Swallowing ImportError would silently skip required parameter restoration and continue with invalid layouts.

File: verl/utils/vllm/vllm_dsv4_fp8_utils.py [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+def reload_deepseek_v4_dense_fp8_scales(model):
+    cache = getattr(model, "_verl_dense_fp8_scale_cache", None)
+    if not cache:
+        return
+
+    params = dict(model.named_parameters())
+    for name, scale in cache.items():
+        param = params.get(name)
+        if param is not None:
+            _copy_scale_shard(param, scale)
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/utils/vllm/vllm_dsv4_fp8_utils.py:352]: These dense scales are explicitly reconstructed as BlockQuantScaleParameter objects. There is no exercised buffer representation. The suggested replacement also removes the empty-cache and missing-name guards, causing errors on normal FP8 models.

**Review by wuxibin89 — APPROVED** [review]

File: verl/utils/vllm/vllm_dsv4_fp8_utils.py [hunk 13]
```diff
@@ -0,0 +1,352 @@
...
+            quant_method = module.quant_method
+            num_experts = quant_method.num_experts
+            intermediate_size = quant_method.intermediate_size
+            hidden_size = quant_method.hidden_size
+            device = module.w13_weight.device
+        else:
+            continue
+
+        weight_loader = module.weight_loader
+        module.w13_weight = _make_mxfp4_moe_param(
```
Comment (wuxibin89) [inline verl/utils/vllm/vllm_dsv4_fp8_utils.py:229]: @aoshen02 Reconstruct weight breaks cuda graph, so we can only run dsv4 with eager mode. Need to support direct weight.copy_: https://github.com/vllm-project/vllm/issues/48312
